src/MolSim.cpp

```diff
@@ -151,121 +167,80 @@ void print_help() {
+  std::cout << "  -f                 Calculate Gravitational Force instead of "
+               "Lennard-Jones Force\n";
+  std::cout << "  -n   Disable all file output for the sake of performance\n";
 }
 
 void calculateF() {
-  for (auto &p1 : particles) {
-    std::array<double, 3> force_copy = p1.getF();
-    for (auto &p2 : particles) {
-      double f_x, f_y, f_z = 0;
-      double distance = std::sqrt(std::pow(p1.getX()[0] - p2.getX()[0], 2) +
-                                  std::pow(p1.getX()[1] - p2.getX()[1], 2) +
-                                  std::pow(p1.getX()[2] - p2.getX()[2], 2));
-      if (!(p1 == p2)) {
-        f_x = (p2.getX()[0] - p1.getX()[0]) * (p1.getM() * p2.getM()) /
-              pow(distance, 3);
-        f_y = (p2.getX()[1] - p1.getX()[1]) * (p1.getM() * p2.getM()) /
-              pow(distance, 3);
-        f_z = (p2.getX()[2] - p1.getX()[2]) * (p1.getM() * p2.getM()) /
-              pow(distance, 3);
-        p1.updateF(p1.getF()[0] + f_x, p1.getF()[1] + f_y, p1.getF()[2] + f_z);
-      }
-    }
-    p1.updateOldF(force_copy[0], force_copy[1], force_copy[2]);
-  }
-}
-
-void calculateF_new() {
-  // Store the current force as the old force and reset current to 0
-  for (auto &p1 : particles) {
-    p1.updateOldF(p1.getF()[0], p1.getF()[1], p1.getF()[2]);
-    p1.updateF(0, 0, 0);
+  // store the current force as the old force and reset current to 0
+  for (auto &p : particles) {
+    p.updateOldF(p.getF());
+    p.updateF(0, 0, 0);
   }
 
-  // Simple force calculation formula (14)
-  for (auto it1 = particles.begin(); it1 != particles.end(); ++it1) {
-    for (auto it2 = ++it1; it2 != particles.end(); ++it2) {
-      Particle &p1 = *it1;
-      Particle &p2 = *it2;
-
-      double distance = std::sqrt(std::pow(p1.getX()[0] - p2.getX()[0], 2) +
-                                  std::pow(p1.getX()[1] - p2.getX()[1], 2) +
-                                  std::pow(p1.getX()[2] - p2.getX()[2], 2));
-
-      // Avoid division by zero
-      if (distance > 0) {
-        double f_x = (p2.getX()[0] - p1.getX()[0]) * (p1.getM() * p2.getM()) /
-                     pow(distance, 3);
-        double f_y = (p2.getX()[1] - p1.getX()[1]) * (p1.getM() * p2.getM()) /
-                     pow(distance, 3);
-        double f_z = (p2.getX()[2] - p1.getX()[2]) * (p1.getM() * p2.getM()) /
-                     pow(distance, 3);
-
-        p1.updateF(p1.getF()[0] + f_x, p1.getF()[1] + f_y, p1.getF()[2] + f_z);
-        // Newton's third law
-        p2.updateF(p2.getF()[0] - f_x, p2.getF()[1] - f_y, p2.getF()[2] - f_z);
+  // Iterate each pair
+  for (auto it = particles.pair_begin(); it != particles.pair_end(); ++it) {
+    ParticlePair &pair = *it;
+    Particle &p1 = *(pair.first);
+    Particle &p2 = *(pair.second);
+    auto r12 = p2.getX() - p1.getX();
+    // distance ||x_i - x_j ||
+    double distance = ArrayUtils::L2Norm(r12);
+
+    // avoid extermely small distance
+    if (distance > 1e-5) {
```

Comment:
This should be removed: If such distances are achieved, it generally suggests that something is wrong in your simulation, for example, a too-large step size, duplicate particles, or something else which is not stable. It is quite dangerous to ignore this. Potentially, you can add a `warn` log message if this happens instead.